hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/RatisPipelineProvider.java

```diff
@@ -98,8 +105,65 @@ private boolean exceedPipelineNumberLimit(ReplicationFactor factor) {
     return false;
   }
 
+  private Map<DatanodeDetails, Integer> getSuggestedLeaderCount(
+      List<DatanodeDetails> dns) {
+    Map<DatanodeDetails, Integer> suggestedLeaderCount = new HashMap<>();
+    for (DatanodeDetails dn : dns) {
+      suggestedLeaderCount.put(dn, 0);
+
+      Set<PipelineID> pipelineIDSet = getNodeManager().getPipelines(dn);
+      for (PipelineID pipelineID : pipelineIDSet) {
+        try {
+          Pipeline pipeline = getPipelineStateManager().getPipeline(pipelineID);
+          if (!pipeline.isClosed()
+              && dn.getUuid().equals(pipeline.getSuggestedLeaderId())) {
```

Comment:
should we use getLeaderId() instead of getSuggestedLeaderId() here to reflect the actual leader count?

Reply:
And the method name can be changed to getLeaderCount() so that the suggest leader is determined by the actual leader count.

Reply:
[RATIS-967](https://issues.apache.org/jira/browse/RATIS-967) guarantee the highpriority node act as leader when create pipeline. In a long running cluster, the leader maybe crash, then some follower will take the leadership, but when the old leader restart and catch up with current leader's log, the old leader can grab the leadership again by [RATIS-967](https://issues.apache.org/jira/browse/RATIS-967). 

So let me suppose the following case, there are 3 servers: s1, s2, s3, there are 2 pipelines now,  the first pipeline's leader is s1, the second pipeline's leader is s2, both the 2 leaders are suggested leader with high priority. Then s1 crash, suppose s3 will take the first pipline's leader.  Then s1 restart, but has not grab leadership of the first pipeline. If we use getLeaderId() instead of getSuggestedLeaderId() to reflect the actual leader count, when we create the 3 third pipeline, we find the leader number on s1, s2, s3 is 0, 1, 1, so we will select s1 as the suggest leader,  then s1 grab the leadership of the first pipeline, so the leader number on s1, s2, s3 will be 2, 1, 0, it's not balance.

Reply:
bq.  then s1 grab the leadership of the first pipeline by RATIS-967,

Does RATIS-967 always gives up its leader when s1 is back online even the current leader works fine? I think this is more specific to RATIS-967 wrt. how the priority is enforced. Any performance impact on the pipeline of forcing leader to be the original one? 

I'm thinking of instead of forcing leader of pipeline P1, P2, P3 like
S1   S2   S3
P1   P2
              P3

In the case of S1 temporarily down, why don't we keep P1 leader on S3 and create P3 with leader on S1, this gives more flexibility for higher level to choose leader?
S1   S2   S3
       P2
              P1
P3

Another situation I'm thinking of is writers on pipeline with slow leader(e.g., hardware slowness) may not be able to recover by leader change.

Reply:
@xiaoyuyao Good point, I also have thought this.

> Any performance impact on the pipeline of forcing leader to be the original one.

If there is performance problem, I can improve forcing leader change within 1 second. I already know how to improve it, but has not implemented it.

> Another situation I'm thinking of is writers on pipeline with slow leader(e.g., hardware slowness) may not be able to recover by leader change.

We can find slow leader by some metric, decrease the priority of the slow leader, select one faster datanode and increase it's priority, so the faster datanode will grab the leadership from the slow leader.


> In the case of S1 temporarily down, why don't we keep P1 leader on S3 and create P3 with leader on S1, this gives more flexibility for higher level to choose leader?

I want the cluster leader distribution as we planned, if the plan is not appropriate, we can adjust the plan by change priority.

If the leader distribution totally depends on hardware rather than plan, we maybe lost control of the leader distribution. Because the leaderId in scm was reported by datanode, it maybe a delayed leaderId. For example, datanode report:

S1  S2  S3
P1 P2

then P1's leader transfer to S3, but SCM has not received this report, SCM allocate P3's leader to S3, then

S1  S2  S3
     P2   P1
            P3.

It's not balance now.

Reply:
Good point, planed leader distribution should work with RATIS-967. How do we scale with this. Plan weight for each of node as a leader when the cluster has thousands of nodes can be difficult.

Reply:
> Plan weight for each of node as a leader when the cluster has thousands of nodes can be difficult.

If each node has similar hardware, i.e. CPU, memory, we just plan weight as now, assign each node with same leader number, it is cheap and reasonable.

I think the only case we need to consider is that some nodes' hardware is weaker than other nodes' obviously.  I think the weeker datanodes should engage in less pipeline than the stronger datanodes,  but ozone does not support this now.
If we can support this, the maxum leader number of each datanode should be less or equal to ((1/3) * the pipeline number it engaged in),  and we select the datanode as the leader which has lowest value of (leader number / pipeline number it engaged in) in 3 datanodes,  this is also cheap.